### PR TITLE
Check $copyProgress is not null before using it in strpos on PHP 8.1

### LINKED ISSUE #320: BlobRestProxy::saveBlobToFile fails on PHP 8.1 with Passing null to parameter #1 is deprecated
### Which service(blob, file, queue, table) does this issue concern?
blob

### Which version of the SDK was used?
1.5.3 (current)

### What's the PHP/OS version?
PHP 8.1

### What problem was encountered?
When calling `BlobRestProxy->saveBlobToFile`, a wild
```
strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated
```
appears

### Steps to reproduce the issue?
Call `BlobRestProxy->saveBlobToFile()`

### Have you found a mitigation/solution?
Yes, as usual :-)

### Is there a failing request ID related to this problem returned by server? What is it?
N/A

### What is the storage account name and time frame of your last reproduce? (UTC YYYY/MM/DD hh:mm:ss)
N/A

(If you think some of the information should not be shared publicly, you can e-mail the main Microsoft contributors of the repository instead.)

### PR DESCRIPTION
**Pull request by spaze**
Passing nulls to string params is deprecated in PHP 8.1.

Fix #320

### CODE REVIEW

**Review by yochanan-gurovich — APPROVED** [review]

**Comment by pimjansen** [comment]
Any update on this? PHP 8.1 is release almost a month ago as stable and this is still blocking the use of it

**Review by pimjansen — APPROVED** [review]

**Comment by spaze** [comment]
Unfortunately, releases supporting new major or minor PHP versions can take very long.

Take for example the previous two major/minor PHP versions:
- 7.4 released in Nov 2019, this lib supports 7.4 since August 2020 release (see e.g. #207)
- PHP 8.0 released in Nov 2020, this lib supports PHP 8.0 starting with September 2021 release (see e.g. #274)

After the 7.4 support took 9 months, the maintainers have promised [quarterly releases](https://github.com/Azure/azure-storage-php/issues/207#issuecomment-658531592) but there never was a quarterly release since then. I can only suggest local modifications using the [composer-patches](https://github.com/cweagans/composer-patches) plugin.

This is what we have done to have this fix (and others too) without needing to wait for a release. I've done this for 8.0 support, and I'm now running a patched version with PHP 8.1.

1. Added extra config to `composer.json`:
```json
	"extra": {
		"patches-file": "patches/composer.patches.json"
	}
```

`patches/composer.patches.json` contains the following:
```json
{
	"patches": {
		"microsoft/azure-storage-blob": {
			"Check null before passing to strpos": "patches/azure-321-check-null.diff"
		}
	}
}
```

You can create the `patches/azure-321-check-null.diff` file by appending `.diff` to the URL of this PR like this https://github.com/Azure/azure-storage-php/pull/321.diff and saving the generated file. Good luck.

**Comment by pimjansen** [comment]
Yeah guess we will fork this then until this gets update. However we use the league adapter package which has a dependency here so its a pain in the ass.

Pretty bad from Microsoft side to maintain this so badly

**Comment by spaze** [comment]
> Yeah guess we will fork this then until this gets update. However we use the league adapter package which has a dependency here so its a pain in the ass.

Either patching or forking (using the composer [`replace`](https://getcomposer.org/doc/04-schema.md#replace) mapping) is way easier and faster than waiting for a release :-)

I personally prefer patching as it's also a bit easier to revert to the original package when released, at least for me. Otherwise I'd have to use three forked packages so I could start using PHP 8.1.

**Comment by pimjansen** [comment]
@katmsft @XiaoningLiu can any of you have a look? Its a shame that Microsoft SDK is still not compatible with PHP.

It is even against the readme where 7.x+ is stated

**Review by pimjansen — APPROVED** [review]

**Review by jaykay-design — APPROVED** [review]

**Comment by SamMousa** [comment]
A workaround, more broadly for 3rd party code is this:
```php
error_reporting(E_ALL & ~E_DEPRECATED);
```
It will still break on the next major PHP version though..

**Comment by nuil** [comment]
So, any update?

**Comment by SamMousa** [comment]
Move away from Azure is the best approach I reckon.
The API SDK is not really being worked on, I'm seeing incidental response times > 3 sec on some of the calls...

**Comment by pimjansen** [comment]
I think it is time to fork. Will we do that on OSS bases in new organisation or what do you guys want?

**Comment by SamMousa** [comment]
I'd say do it in a new organisation whether or not it becomes a real community effort. That way you're ready for the future. I myself will probably not have a lot of time to contribute.

**Review by M4SX5 — APPROVED** [review]
Is working on PHP 8.2

**Comment by SamMousa** [comment]
@M4SX5 is this a sign that there will be actual maintenance on this package?

**Comment by kyleweishaupt** [comment]
Just ran into this issue after migrating our app to Docker & PHP 8.1. Using the League adapter as well. Hope this gets merged soon.

**Comment by spaze** [comment]
> Hope this gets merged soon.

Hopefully but please don't hold your breath, it's 9 months already after all! I'd suggest following this guide https://github.com/Azure/azure-storage-php/pull/321#issuecomment-991373427 and patch your lib.

**Comment by pimjansen** [comment]
@spaze much wow!

**Pull request merged**
